### ISSUE TITLE
Optimistic lock problem

### ISSUE THREAD

**Issue by AlexGx** [yii2tech/ar-softdelete#13]
Optimistic lock logic is not taken into account when soft delete/restore

my quick fix
replace in delete/restore internal methods
```php
$result = $this->owner->updateAttributes($attributes);
```
with
```php
$this->owner->setAttributes($attributes, false);
$result = $this->owner->update(false, array_keys($attributes));
```

**Comment by klimov-paul** [comment]
No, this will cause infinite loop around `BaseActiveRecord::EVENT_BEFORE_UPDATE` triggering.

**Comment by tunecino** [comment]
I do it at model setup. It is a bit hacky but at least for now and until https://github.com/yiisoft/yii2/issues/9133 get fixed, it works.

```php
public function behaviors()
{
    return [
        'softDeleteBehavior' => [
            'class' => SoftDeleteBehavior::className(),
            'invokeDeleteEvents' => false,
            'softDeleteAttributeValues' => [
                'isDeleted' => function() {
                    $lock = $this->optimisticLock();
                    $input = Yii::$app->getRequest()->getBodyParam($lock);
                    $isValid = $input && (new NumberValidator())->validate($input);
                    $input = $isValid ? $input : 0;
                    if ($input != $this->version) {
                        throw new StaleObjectException('The object being updated is outdated.');
                    }
                    return true;
                },
            ],
        ],
    ];
}
```

**Comment by klimov-paul** [comment]
Resolved by commit b5f8aae484fd8c6f1254fe3dfd60a671ab2f7cc8
